src/lua/httpc.c

```diff
@@ -607,7 +615,11 @@ luaT_httpc_io_finish(lua_State *L)
 static int
 luaT_httpc_io_cleanup(lua_State *L)
 {
-	httpc_io_destroy(luaT_httpc_checkio(L));
+	struct httpc_io *io = luaT_httpc_checkio(L);
+	struct httpc_env *env = io->req->env;
+	httpc_io_destroy(io);
+	if (env->req_count == 0)
+		luaL_unref(L, LUA_REGISTRYINDEX, env->lua_ref);
```

Comment:
To be honest it's not clear why did you add one lua_ref but two lua_unref. There is a risk that double free (in Lua world) will happen.

Reply:
> To be honest it's not clear why did you add one lua_ref but two lua_unref. 

First case: there is a http client object and zero http chunked requests. We need to call `luaL_unref` in http client obj GC finalizer (`luaT_httpc_cleanup`), otherwise `env->lua_ref` will not be unreferenced.

Second case: there is a http client with a number of chunked http requests. We need to call  `luaL_unref` in the latest http chunked request GC finalizer (`luaT_httpc_io_cleanup`).

> There is a risk that double free (in Lua world) will happen.

Right, it is not good.

I propose another way to link httpc and http req Lua objects:

```diff
diff --git a/src/httpc.h b/src/httpc.h
index c78dbf6e90..bd9f498d9a 100644
--- a/src/httpc.h
+++ b/src/httpc.h
@@ -70,8 +70,6 @@ struct httpc_env {
 	struct mempool req_pool;
 	/** Statistics */
 	struct httpc_stat stat;
-	/** Lua reference. */
-	int lua_ref;
 };
 
 /**
diff --git a/src/lua/httpc.c b/src/lua/httpc.c
index 1679e8c7d7..743442a314 100644
--- a/src/lua/httpc.c
+++ b/src/lua/httpc.c
@@ -495,9 +495,6 @@ luaT_httpc_new(lua_State *L)
 	if (httpc_env_create(ctx, max_conns, max_total_conns) != 0)
 		return luaT_error(L);
 
-	lua_pushvalue(L, -1); /* Popped by luaL_ref(). */
-	ctx->lua_ref = luaL_ref(L, LUA_REGISTRYINDEX);
-
 	luaL_getmetatable(L, DRIVER_LUA_UDATA_NAME);
 	lua_setmetatable(L, -2);
 
@@ -511,7 +508,6 @@ luaT_httpc_cleanup(lua_State *L)
 	httpc_env_finish(env);
 	if (env->req_count == 0) {
 		httpc_env_destroy(env);
-		luaL_unref(L, LUA_REGISTRYINDEX, env->lua_ref);
 	}
 
 	/** remove all methods operating on ctx */
@@ -616,10 +612,7 @@ static int
 luaT_httpc_io_cleanup(lua_State *L)
 {
 	struct httpc_io *io = luaT_httpc_checkio(L);
-	struct httpc_env *env = io->req->env;
 	httpc_io_destroy(io);
-	if (env->req_count == 0)
-		luaL_unref(L, LUA_REGISTRYINDEX, env->lua_ref);
 
 	/** remove all methods operating on io */
 	lua_newtable(L);
diff --git a/src/lua/httpc.lua b/src/lua/httpc.lua
index 3fc5fe4ee6..32f0917d6e 100644
--- a/src/lua/httpc.lua
+++ b/src/lua/httpc.lua
@@ -616,6 +616,7 @@ curl_mt = {
                     },
                     _internal = {
                         io = resp._internal.io,
+                        httpc = self,
                     },
                 })
             end
```